File: relayd/internal/agent/executor.go
package agent

import (
	"archive/tar"
	"compress/gzip"
	"context"
	"fmt"
	"io"
	"os"
	"os/exec"
	pathpkg "path"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	relayv1 "github.com/matteomarolt/relay/relayd/internal/proto/relayv1"
)

// Events is how a running job reports back into the session loop.
type Events interface {
	Status(runID string, state relayv1.RunState, detail string, exitCode int, resultSha string)
	StatusEndpoint(runID string, state relayv1.RunState, detail, endpoint string)
	Log(runID, line string, stderr bool)
	// FetchBlob writes the blob to path; UploadFile returns its sha256.
	FetchBlob(ctx context.Context, sha, path string) error
	UploadFile(ctx context.Context, path string) (string, error)
	// Secrets resolves declared secret names to values at run start.
	Secrets(ctx context.Context, runID string, names []string) (map[string]string, error)
}

// userErrorExit mirrors relay.runtime.invoke.USER_ERROR_EXIT.
const userErrorExit = 13

// sdkBundleDir is where the SDK embeds its own source inside every code
// bundle so containers need no Relay preinstalled. Mirrored in the Python
// client (bundle.py).
const sdkBundleDir = "_relay_sdk"

var volumeNameRe = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._\-]*$`)

func validVolumeName(name string) bool {
	return volumeNameRe.MatchString(name)
}

// reservedMountConflict returns a reason when a CLEANED absolute mount path
// would shadow Relay's runtime paths ("" = fine). Blocks equality, parents
// of reserved paths, and anything inside the io/runtime trees; nested
// mounts inside /workspace are allowed (that's what volumes are for).
func reservedMountConflict(clean string) string {
	reserved := []string{"/relay/io", "/workspace", "/opt/relay"}
	if clean == "/" {
		return "would shadow the filesystem root"
	}
	for _, r := range reserved {
		if clean == r {
			return "collides with the Relay-reserved path " + r
		}
		if strings.HasPrefix(r+"/", clean+"/") { // clean is a parent of r
			return "contains the Relay-reserved path " + r
		}
	}
	for _, tree := range []string{"/relay", "/opt/relay"} {
		if strings.HasPrefix(clean+"/", tree+"/") {
			return "is inside the Relay-reserved tree " + tree
		}
	}
	return ""
}

// DockerExecutor runs RunSpecs in containers. The contract matches the M0
// local executor exactly (io dir + invoke protocol); only the transport
// around it changed. Jobs run once; services (service.go) are supervised.
type DockerExecutor struct {
	docker   *dockerClient
	workRoot string
}

func NewDockerExecutor(docker *dockerClient, workRoot string) *DockerExecutor {
	return &DockerExecutor{docker: docker, workRoot: workRoot}
}

type runPaths struct {
	runDir    string
	workspace string
	ioDir     string
}

// prepare materializes bundle + call payload and ensures the image exists,
// reporting failures itself. ok=false means a status was already sent.
func (e *DockerExecutor) prepare(ctx context.Context, spec *relayv1.RunSpec, ev Events) (runPaths, bool) {
	runID := spec.GetRunId()
	fail := func(format string, args ...any) {
		ev.Status(runID, relayv1.RunState_RUN_ERROR, fmt.Sprintf(format, args...), -1, "")
	}
	p := runPaths{runDir: filepath.Join(e.workRoot, "runs", runID)}
	p.workspace = filepath.Join(p.runDir, "workspace")
	p.ioDir = filepath.Join(p.runDir, "io")
	for _, dir := range []string{p.workspace, p.ioDir} {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			fail("prepare run dir: %v", err)
			return p, false
		}
	}

	ev.Status(runID, relayv1.RunState_RUN_BUILDING, "fetching code bundle", 0, "")
	bundlePath := filepath.Join(p.runDir, "bundle.tar.gz")
	if err := ev.FetchBlob(ctx, spec.GetBundleSha256(), bundlePath); err != nil {
		fail("fetch bundle: %v", err)
		return p, false
	}
	if err := extractTarGz(bundlePath, p.workspace); err != nil {
		fail("extract bundle: %v", err)
		return p, false
	}
	if err := ev.FetchBlob(ctx, spec.GetArgsSha256(),
		filepath.Join(p.ioDir, "call.bin")); err != nil {
		fail("fetch call payload: %v", err)
		return p, false
	}

	present, err := e.docker.ImageExists(ctx, spec.GetImageTag())
	if err != nil {
		fail("docker unavailable: %v", err)
		return p, false
	}
	if !present {
		ev.Status(runID, relayv1.RunState_RUN_BUILDING,
			"building image "+spec.GetImageTag(), 0, "")
		if err := e.buildImage(ctx, spec, ev); err != nil {
			fail("image build failed: %v", err)
			return p, false
		}
	}
	return p, true
}

// baseConfig assembles the shared container configuration (env incl.
// secrets, mounts incl. volumes, resource limits, GPU pinning).
func (e *DockerExecutor) baseConfig(ctx context.Context, spec *relayv1.RunSpec, p runPaths, ev Events) (*containerConfig, error) {
	runID := spec.GetRunId()
	env := []string{
		"PYTHONPATH=/workspace/" + sdkBundleDir,
		"RELAY_RUN_ID=" + runID,
	}
	if names := spec.GetSecretNames(); len(names) > 0 {
		values, err := ev.Secrets(ctx, runID, names)
		if err != nil {
			return nil, fmt.Errorf("resolve secrets: %w", err)
		}
		for name, value := range values {
			env = append(env, name+"="+value)
		}
	}

	cfg := &containerConfig{
		Image:      spec.GetImageTag(),
		Entrypoint: []string{"python3"}, // registry images keep their own otherwise
		Cmd:        []string{"-m", "relay.runtime.invoke", "/relay/io"},
		// Start OUTSIDE /workspace: `python3 -m` prepends the cwd to
		// sys.path, and a project shipping its own relay.py must not shadow
		// the embedded runtime. invoke chdirs to /workspace before user code.
		WorkingDir: "/relay/io",
		Env:        env,
		Labels:     map[string]string{"dev.relay.run": runID},
	}
	// The workspace is a per-run scratch copy, so it mounts read-write.
	// unlike M0 local mode, which mounts the user's real project read-only.
	cfg.HostConfig.Binds = []string{
		p.ioDir + ":/relay/io",
		p.workspace + ":/workspace",
	}
	for mountPath, volName := range spec.GetVolumes() {
		clean := pathpkg.Clean(mountPath)
		if !validVolumeName(volName) || !pathpkg.IsAbs(clean) {
			return nil, fmt.Errorf("invalid volume mapping %q → %q", mountPath, volName)
		}
		// Canonicalized reserved-path guard: no mount may equal, contain,
		// or live inside the runtime's own paths.
		if reason := reservedMountConflict(clean); reason != "" {
			return nil, fmt.Errorf("volume mount %q %s", mountPath, reason)
		}
		volDir := filepath.Join(e.workRoot, "volumes", volName)
		if err := os.MkdirAll(volDir, 0o755); err != nil {
			return nil, fmt.Errorf("create volume %s: %w", volName, err)
		}
		cfg.HostConfig.Binds = append(cfg.HostConfig.Binds, volDir+":"+clean)
	}
	if res := spec.GetResources(); res != nil {
		if res.GetCpus() > 0 {
			cfg.HostConfig.NanoCpus = int64(res.GetCpus() * 1e9)
		}
		if res.GetMemoryMib() > 0 {
			cfg.HostConfig.Memory = int64(res.GetMemoryMib()) << 20
		}
	}
	// Pin to exactly the devices the scheduler granted, never "all".
	if spec.GetAcceleratorKind() == "cuda" && len(spec.GetDeviceIndices()) > 0 {
		ids := make([]string, 0, len(spec.GetDeviceIndices()))
		for _, idx := range spec.GetDeviceIndices() {
			ids = append(ids, fmt.Sprintf("%d", idx))
		}
		cfg.HostConfig.DeviceRequests = []deviceRequest{{
			Driver: "nvidia", DeviceIDs: ids, Capabilities: [][]string{{"gpu"}},
		}}
	}
	return cfg, nil
}

// cleanupContainer force-removes with a bound so a wedged daemon can't pin
// the run goroutine forever.
func (e *DockerExecutor) cleanupContainer(ctx context.Context, runID, containerID string, ev Events) {
	cctx, cancel := context.WithTimeout(context.WithoutCancel(ctx), 30*time.Second)
	defer cancel()
	if err := e.docker.ContainerRemove(cctx, containerID); err != nil {
		ev.Log(runID, "[relay] container cleanup failed: "+err.Error(), true)
	}
}

// Run executes a one-shot job.
func (e *DockerExecutor) Run(ctx context.Context, spec *relayv1.RunSpec, ev Events) {
	runID := spec.GetRunId()
	fail := func(format string, args ...any) {
		ev.Status(runID, relayv1.RunState_RUN_ERROR, fmt.Sprintf(format, args...), -1, "")
	}

	p, ok := e.prepare(ctx, spec, ev)
	if !ok {
		return
	}
	defer os.RemoveAll(p.runDir)

	cfg, err := e.baseConfig(ctx, spec, p, ev)
	if err != nil {
		fail("%v", err)
		return
	}
	containerID, err := e.docker.ContainerCreate(ctx, "relay-"+runID, cfg)
	if err != nil {
		fail("create container: %v", err)
		return
	}
	defer e.cleanupContainer(ctx, runID, containerID, ev)

	if err := e.docker.ContainerStart(ctx, containerID); err != nil {
		fail("start container: %v", err)
		return
	}
	ev.Status(runID, relayv1.RunState_RUN_RUNNING, "", 0, "")

	// The log follower gets its own cancelable context so no failure mode
	// (wedged daemon, failed removal) can strand this goroutine forever.
	logCtx, logCancel := context.WithCancel(context.WithoutCancel(ctx))
	defer logCancel()
	logsDone := make(chan struct{})
	go func() {
		defer close(logsDone)
		_ = e.docker.ContainerLogs(logCtx, containerID, func(line string, stderr bool) {
			ev.Log(runID, line, stderr)
		})
	}()
	joinLogs := func() {
		select {
		case <-logsDone:
		case <-time.After(30 * time.Second):
			logCancel()
			<-logsDone
		}
	}

	timeout := time.Duration(spec.GetResources().GetTimeoutS()) * time.Second
	exitCode, timedOut, err := e.docker.WaitWithTimeout(ctx, containerID, timeout)
	if timedOut || err != nil {
		// Kill the container first because the log follower only ends when the
		// container does, so joining logs before removal deadlocks.
		e.cleanupContainer(ctx, runID, containerID, ev)
		joinLogs()
		if timedOut {
			fail("run exceeded its %s timeout and was killed", timeout)
			return
		}
		if ctx.Err() != nil { // canceled from the server
			ev.Status(runID, relayv1.RunState_RUN_CANCELED, "canceled", -1, "")
			return
		}
		fail("wait: %v", err)
		return
	}
	joinLogs()

	// Collect the result.
	resultPath := filepath.Join(p.ioDir, "result.bin")
	if _, statErr := os.Stat(resultPath); statErr != nil ||
		(exitCode != 0 && exitCode != userErrorExit) {
		fail("container exited %d without a result. This is an infrastructure failure; "+
			"see run logs", exitCode)
		return
	}
	resultSha, err := uploadResultWithRetry(ctx, ev, runID, resultPath)
	if err != nil {
		fail("upload result: %v", err)
		return
	}
	if exitCode == userErrorExit {
		ev.Status(runID, relayv1.RunState_RUN_FAILED,
			"function raised", exitCode, resultSha)
	} else {
		ev.Status(runID, relayv1.RunState_RUN_SUCCEEDED, "", 0, resultSha)
	}
}

// uploadResultWithRetry is PATIENT: a result that finished while the
// control plane was offline (dev laptop asleep, server rebooting) must
// survive until the server returns. The run stays active agent-side, so
// reconcile keeps it alive and the terminal status flushes after upload.
func uploadResultWithRetry(ctx context.Context, ev Events, runID, path string) (string, error) {
	backoff := 5 * time.Second
	deadline := time.Now().Add(24 * time.Hour)
	for attempt := 0; ; attempt++ {
		sha, err := ev.UploadFile(ctx, path)
		if err == nil {
			if attempt > 0 {
				ev.Log(runID, "[relay] result uploaded after reconnect", true)
			}
			return sha, nil
		}
		if ctx.Err() != nil || time.Now().After(deadline) {
			return "", err
		}
		if attempt == 0 {
			ev.Log(runID, "[relay] result ready but the server is "+
				"unreachable. Retrying until it returns", true)
		}
		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case <-time.After(backoff):
		}
		if backoff < time.Minute {
			backoff *= 2
		}
	}
}

// buildImage shells out to the docker CLI: BuildKit access without linking
// BuildKit. Build output streams into the run's logs.
func (e *DockerExecutor) buildImage(ctx context.Context, spec *relayv1.RunSpec, ev Events) error {
	cmd := exec.CommandContext(ctx, "docker", "build",
		"--progress=plain", "-t", spec.GetImageTag(), "-")
	cmd.Stdin = strings.NewReader(spec.GetImageDockerfile())
	out, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	cmd.Stderr = cmd.Stdout
	if err := cmd.Start(); err != nil {
		return fmt.Errorf("docker CLI not available for image build: %w", err)
	}
	sc := newLineScanner(out)
	for sc.Scan() {
		ev.Log(spec.GetRunId(), "[build] "+sc.Text(), true)
	}
	return cmd.Wait()
}

// extractTarGz unpacks a bundle, refusing entries that escape the target.
func extractTarGz(archive, dest string) error {
	f, err := os.Open(archive)
	if err != nil {
		return err
	}
	defer f.Close()
	gz, err := gzip.NewReader(f)
	if err != nil {
		return err
	}
	defer gz.Close()
	tr := tar.NewReader(gz)
	destAbs, err := filepath.Abs(dest)
	if err != nil {
		return err
	}
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		target := filepath.Join(destAbs, hdr.Name)
		if !strings.HasPrefix(target, destAbs+string(os.PathSeparator)) && target != destAbs {
			return fmt.Errorf("bundle entry %q escapes the workspace", hdr.Name)
		}
		switch hdr.Typeflag {
		case tar.TypeDir:
			if err := os.MkdirAll(target, 0o755); err != nil {
				return err
			}
		case tar.TypeReg:
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				return err
			}
			w, err := os.OpenFile(target, os.O_CREATE|os.O_WRONLY|os.O_TRUNC,
				os.FileMode(hdr.Mode)&0o777|0o400)
			if err != nil {
				return err
			}
			if _, err := io.Copy(w, tr); err != nil {
				w.Close()
				return err
			}
			w.Close()
		case tar.TypeSymlink, tar.TypeLink:
			// Links could point outside the workspace; bundles never
			// legitimately contain them (the SDK skips them at pack time).
			return fmt.Errorf("bundle entry %q: links are not allowed", hdr.Name)
		}
	}
}

File: sdk/src/relay/_local/executor.py
"""M0/M1 embedded local executor: run one decorated function in a local
Docker container, stream its logs, and return its result.

Deliberately the smallest thing that makes the DX real. The fleet path
(relayd agent, Go) implements the same io-dir + invoke protocol; only the
transport differs.

Timeouts are enforced by an in-process watchdog that force-removes the
container at the deadline for both `.remote()` and `.spawn()` (the
watchdog lives as long as this Python process; the fleet path enforces
timeouts server-side regardless of any client).
"""

from __future__ import annotations

import json
import secrets
import sys
import threading
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator

import relay

from ..exceptions import ExecutorError, RemoteFunctionError
from ..image import Image
from ..resources import Resources
from ..runtime import serialize
from ..runtime.invoke import USER_ERROR_EXIT
from . import docker

RUNS_DIR = Path.home() / ".relay" / "runs"

_UNSET = object()


@dataclass(frozen=True)
class CallSpec:
    function_name: str
    entry_module: str
    project_dir: Path
    image: Image
    resources: Resources
    args: tuple[Any, ...]
    kwargs: dict[str, Any]
    app_name: str = "app"
    target_names: tuple[str, ...] | list[str] = ()
    volumes: dict[str, str] | None = None  # mount path → volume name
    secrets: list[str] | None = None       # env-var names


def _new_run_id() -> str:
    return f"run_{int(time.time()):x}{secrets.token_hex(3)}"


def _relay_pkg_root() -> Path:
    """Directory containing the `relay` package, mounted at /opt/relay so
    containers need no Relay install."""
    return Path(relay.__file__).resolve().parent.parent


def _docker_args(spec: CallSpec, run_id: str, io_dir: Path) -> list[str]:
    args = [
        "--name", f"relay-{run_id}",
        "--entrypoint", "python3",  # registry images keep their own otherwise
        "-v", f"{io_dir}:/relay/io",
        "-v", f"{spec.project_dir.resolve()}:/workspace:ro",
        "-v", f"{_relay_pkg_root()}:/opt/relay:ro",
        "-e", "PYTHONPATH=/opt/relay",
        "-e", f"RELAY_RUN_ID={run_id}",
        # cwd starts outside /workspace (shadowing guard); invoke chdirs in.
        "--workdir", "/relay/io",
    ]
    for mount, name in (spec.volumes or {}).items():
        vol_dir = Path.home() / ".relay" / "volumes" / name
        vol_dir.mkdir(parents=True, exist_ok=True)
        args += ["-v", f"{vol_dir}:{mount}"]
    # Local mode resolves secrets from the client's own environment.
    import os as _os

    for name in spec.secrets or []:
        if name not in _os.environ:
            raise ExecutorError(
                f"{spec.function_name} declares Secret({name!r}) but local "
                f"mode reads it from YOUR environment. Export {name}=... "
                f"(fleet mode uses `relay secret set {name}` instead)."
            )
        args += ["-e", f"{name}={_os.environ[name]}"]
    if spec.resources.cpu:
        args += [f"--cpus={spec.resources.cpu:g}"]
    if spec.resources.memory_mib:
        args += [f"--memory={spec.resources.memory_mib}m"]

    wants_cuda = spec.resources.accelerators is not None and any(
        acc.kind == "cuda" for acc in spec.resources.accelerators.options
    )
    if wants_cuda:
        if docker.host_nvidia_available():
            args += ["--gpus", "all"]
        else:
            print(
                f"⚠ relay: {spec.function_name} asks for "
                f"{spec.resources.accelerators.describe()}, but LOCAL mode "
                f"on this machine has no CUDA. Running without a GPU. "
                f"(Local mode never enforces resource specs; the fleet "
                f"scheduler does.)",
                file=sys.stderr,
            )
    args += [spec.image.tag(), "-m", "relay.runtime.invoke", "/relay/io"]
    return args


def _ensure_image(image: Image) -> None:
    tag = image.tag()
    if docker.image_exists(tag):
        return
    print(f"→ building image {tag} (first run for this definition)", file=sys.stderr)
    docker.build_image(tag, image.dockerfile(), on_line=lambda l: print(f"  {l}", file=sys.stderr))
    print(f"✓ image {tag} ready", file=sys.stderr)


class _Watchdog:
    """Force-removes the container when the deadline passes."""

    def __init__(self, container: str, timeout_s: int | None):
        self.fired = False
        self._timer: threading.Timer | None = None
        if timeout_s:
            def fire() -> None:
                self.fired = True
                docker.force_remove(container)

            self._timer = threading.Timer(timeout_s, fire)
            self._timer.daemon = True
            self._timer.start()

    def cancel(self) -> None:
        if self._timer is not None:
            self._timer.cancel()


def _read_result(io_dir: Path, exit_code: int, run_id: str) -> Any:
    result_path = io_dir / serialize.RESULT_FILENAME
    if exit_code in (0, USER_ERROR_EXIT) and result_path.exists():
        result = serialize.load_result(result_path)
        if result["ok"]:
            return result["value"]
        raise RemoteFunctionError(
            result["exc_type"], result["message"], result["traceback"]
        )
    raise ExecutorError(
        f"Run {run_id} exited with code {exit_code} before producing a "
        f"result. This is a runtime failure, not your function raising. "
        f"Check the log output above; io dir kept at {io_dir.parent}."
    )


class LocalRun:
    """Handle for a spawned local run. Mirrors the fleet Run API surface.

    Terminal state is persisted to meta.json, so status()/result() keep
    answering after the container is gone.
    """

    def __init__(self, run_id: str, container_id: str, io_dir: Path,
                 spec: CallSpec, watchdog: _Watchdog):
        self.run_id = run_id
        self._container = container_id
        self._io_dir = io_dir
        self._spec = spec
        self._watchdog = watchdog
        self._result: Any = _UNSET

    # -- metadata ---------------------------------------------------------

    @property
    def _meta_path(self) -> Path:
        return self._io_dir.parent / "meta.json"

    def _load_meta(self) -> dict:
        try:
            return json.loads(self._meta_path.read_text())
        except (OSError, json.JSONDecodeError):
            return {}

    def _save_meta(self, **fields: Any) -> None:
        meta = self._load_meta()
        meta.update(fields)
        self._meta_path.write_text(json.dumps(meta))

    # -- public surface ---------------------------------------------------

    def status(self) -> str:
        if state := self._load_meta().get("state"):
            return state
        state = docker.inspect_state(self._container)
        if not state:
            return "unknown"
        if state.get("Running"):
            return "running"
        code = state.get("ExitCode", -1)
        if code == 0:
            return "succeeded"
        if code == USER_ERROR_EXIT:
            return "failed"
        return f"error({code})"

    def logs(self, follow: bool = True) -> Iterator[str]:
        return docker.stream_logs(self._container, follow=follow)

    def result(self, timeout: float | None = None) -> Any:
        """Wait for and return the function's result.

        `timeout` bounds the WAIT, not the run: on expiry the run keeps
        going and this raises TimeoutError (cancel() to kill it). The run's
        own `timeout=` from the decorator is enforced by the watchdog.
        """
        if self._result is not _UNSET:
            return self._result
        exit_code = docker.wait(self._container, timeout_s=timeout)
        self._watchdog.cancel()
        if self._watchdog.fired:
            self._save_meta(state="error", detail="timeout")
            raise ExecutorError(
                f"Run {self.run_id} exceeded its "
                f"{self._spec.resources.timeout_s}s timeout and was killed."
            )
        try:
            value = _read_result(self._io_dir, exit_code, self.run_id)
        except RemoteFunctionError:
            self._save_meta(state="failed", exit_code=exit_code)
            docker.force_remove(self._container)
            raise
        except ExecutorError:
            self._save_meta(state="error", exit_code=exit_code)
            docker.force_remove(self._container)
            raise
        self._save_meta(state="succeeded", exit_code=0)
        docker.force_remove(self._container)
        self._result = value
        return value

    def cancel(self) -> None:
        self._watchdog.cancel()
        self._save_meta(state="canceled")
        docker.force_remove(self._container)


class LocalExecutor:
    def _prepare(self, spec: CallSpec) -> tuple[str, Path]:
        docker.ensure_docker()
        _ensure_image(spec.image)
        run_id = _new_run_id()
        io_dir = RUNS_DIR / run_id / "io"
        io_dir.mkdir(parents=True, exist_ok=True)
        serialize.dump_call(
            io_dir / serialize.CALL_FILENAME,
            entry_module=spec.entry_module,
            function=spec.function_name,
            args=spec.args,
            kwargs=spec.kwargs,
        )
        return run_id, io_dir

    def run_sync(self, spec: CallSpec) -> Any:
        run_id, io_dir = self._prepare(spec)
        print(
            f"→ {spec.function_name} on local-docker "
            f"[{spec.resources.describe()}] ({run_id})",
            file=sys.stderr,
        )
        container = docker.run_detached(_docker_args(spec, run_id, io_dir))
        watchdog = _Watchdog(container, spec.resources.timeout_s)
        try:
            for line in docker.stream_logs(container):
                print(line)
            exit_code = docker.wait(container, timeout_s=60)
            watchdog.cancel()
            if watchdog.fired:
                raise ExecutorError(
                    f"Run {run_id} exceeded its "
                    f"{spec.resources.timeout_s}s timeout and was killed."
                )
            return _read_result(io_dir, exit_code, run_id)
        finally:
            watchdog.cancel()
            docker.force_remove(container)

    def spawn(self, spec: CallSpec) -> LocalRun:
        run_id, io_dir = self._prepare(spec)
        container = docker.run_detached(_docker_args(spec, run_id, io_dir))
        watchdog = _Watchdog(container, spec.resources.timeout_s)
        print(f"→ spawned {spec.function_name} ({run_id})", file=sys.stderr)
        return LocalRun(run_id, container, io_dir, spec, watchdog)
